Support IPC for Expandable Segments

zdevito opened the pull request:
This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130890

Differential Revision: [D60547506](https://our.internmc.facebook.com/intern/diff/D60547506)

zdevito committed:
Support IPC for Expandable Segments

[ghstack-poisoned]

zdevito committed:
Update on "Support IPC for Expandable Segments"




[ghstack-poisoned]

zdevito committed:
Update on "Support IPC for Expandable Segments"




[ghstack-poisoned]

zdevito committed:
Update on "Support IPC for Expandable Segments"




[ghstack-poisoned]

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -420,6 +426,7 @@ struct ExpandableSegment {
       CUmemGenericAllocationHandle handle = 0;
       CUmemAllocationProp prop = {};
       prop.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+      prop.requestedHandleTypes = CU_MEM_HANDLE_TYPE_POSIX_FILE_DESCRIPTOR;
```
I ask some Nvidia folks and they verified that this doesn't consume additional resources if we do not request a handle. So it is safe to do this for all expandable segments.

ngimel replied:
For older drivers (like 525) this sometimes leads to big slowdowns.

zdevito committed:
Update on "Support IPC for Expandable Segments"




[ghstack-poisoned]

dsjohns2 commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -3386,42 +3527,78 @@ class NativeCachingAllocator : public CUDAAllocator {
   // it will deleted in cudaIpcCloseMemHandle when its reference count is 0.
```
I think we have to update some language in this comment block.

dsjohns2 commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -3386,42 +3527,78 @@ class NativeCachingAllocator : public CUDAAllocator {
+      if (cuda_ipc_ptr_) {
+        cuda::CUDAGuard device_guard(device_);
+        C10_CUDA_CHECK(cudaIpcCloseMemHandle(cuda_ipc_ptr_));
+      }
```
Just for my understanding, we don't have to close in the expandable segments case because it is already taken care of in fromShared: close((int)myfd); ?

zdevito replied:
The expandable segment case is handled by the ExpandableSegment object dtor. We only need something special here because cuda_ipc_ptr_ is a raw object.

dsjohns2 reviewed (approved):
I think LGTM overall! A few minor questions/comments.

zdevito committed:
Update on "Support IPC for Expandable Segments"




[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

deepshah133 commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -3511,21 +3511,30 @@ class NativeCachingAllocator : public CUDAAllocator {
+
+  // When using cuMemCreate, via expandable segments, we use c
+  // MemExportToShareableHandle
+  // create a file descriptor that can be sent t
+  //  the other process to
+  // ort the object. Then we recreate part of the ex
+  // andable segment necessary to
+  // the allocation.
+
+  // ipcMemHandle_to_devptr caches the mapping from shareable handle to
```
This comment block seems incomplete.

deepshah133 commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -454,10 +461,77 @@ struct ExpandableSegment {
+      auto& handle = handles_.at(i).value();
+      if (!handle.fd) {
+        int fd = 0;
+        C10_CUDA_DRIVER_CHECK(DriverAPI::get()->cuMemExportToShareableHandle_(
```
Couple of questions for my understanding (pardon my noob understanding)

- Does obtaining exportable handle increment the refcount of the backing handle? (i.e. does it need to be closed explicitly?).
- Are we overwriting the previous fd (original one obtained with cuMemCreate) below with newly obtained fd?

zdevito replied:
The export creates a file descriptor. I do not think the cuMemCreate command actually creates a file descriptor. It only tells the library that one may be requested in the future.

deepshah133 commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -454,10 +461,77 @@ struct ExpandableSegment {
+      (void)i;
+      int fd = 0;
+      buf.read((char*)&fd, sizeof(int));
+      auto myfd = syscall(SYS_pidfd_getfd, pidfd, fd, 0);
```
Not well familiar with pidfd_getfd, but IIUC this create a duplicate fd (referring to same resource) from other process. Is that required? I thought the fd obtained via cuMemExportToShareableHandle can be directly plugged into cuMemImportFromShareableHandle (or are we worried about it being released in-between?)

zdevito replied:
Since they are in different processes, they need to be different file descriptor numbers. The things that are needed are taken from the only same of this API: https://github.com/NVIDIA/cuda-samples/tree/master/Samples/3_CUDA_Features/memMapIPCDrv

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot revert

zdevito commented:
@pytorchbot revert -m 'variable shadowing broke internal tests'

zdevito commented:
@pytorchbot revert -m 'variable shadowing broke internal tests' -c ghfirst

zdevito committed:
Update on "Support IPC for Expandable Segments"




[ghstack-poisoned]

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.




[ghstack-poisoned]

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.




[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.




[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot merge -f "this is a reland of something that only broke a warning-as-error error in an internal build, skipping tests that have been queued overnight"

zdevito commented:
@pytorchbot revert -m "some internal tests show shutdown issues with the change to the table that holds ipc handles"

zdevito commented:
@pytorchbot revert -m "some internal tests show shutdown issues with the change to the table that holds ipc handles" -c "ghfirst"

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.




[ghstack-poisoned]

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60547506).

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.


Differential Revision: [D60547506](https://our.internmc.facebook.com/intern/diff/D60547506)

[ghstack-poisoned]

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.


Differential Revision: [D60547506](https://our.internmc.facebook.com/intern/diff/D60547506)

[ghstack-poisoned]

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.


Differential Revision: [D60547506](https://our.internmc.facebook.com/intern/diff/D60547506)

[ghstack-poisoned]

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.


Differential Revision: [D60547506](https://our.internmc.facebook.com/intern/diff/D60547506)

[ghstack-poisoned]

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60547506).

zdevito committed:
Update on "Support IPC for Expandable Segments"

This reapplication commit is the same as before except it resolves a build error in an internal build where `handle` was shadowed.


Differential Revision: [D60547506](https://our.internmc.facebook.com/intern/diff/D60547506)

[ghstack-poisoned]

ngimel commented:
btw cuda 12.5+ (and corresponding driver) supports CU_MEM_HANDLE_TYPE_FABRIC that can be shared without all the fd exchange pain.

youkaichao commented:
@ngimel @zdevito thanks for the great PR! I tried it in my code with https://gist.github.com/youkaichao/8f87555bdeaaf68f4492b0dc96fbd206 , it seems not working. Can you please help?

youkaichao commented:
running on `2.6.0.dev20241112+cu124` , still get the same error `RuntimeError: pidfd_getfd: Operation not permitted` .

is there any specific compiled version of pytorch i need to install?

I use `pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu124`

AlJohri commented:
@youkaichao were you able to figure out a workaround for the `RuntimeError: pidfd_getfd: Operation not permitted` issue? we are running into the same issue

youkaichao commented:
@AlJohri I think the root cause is somehow your os disables ptrace even if the processes are launched by the same user.

you can try to add the following code to make sure your processes can be ptraced, so that the exported fd can be used by other processes:

```python
import ctypes

# Constants from prctl.h
PR_SET_PTRACER = 0x59616d61
PR_SET_PTRACER_ANY = -1  # Allow any process with the same UID to ptrace

libc = ctypes.CDLL("libc.so.6", use_errno=True)

result = libc.prctl(PR_SET_PTRACER, PR_SET_PTRACER_ANY, 0, 0, 0)
if result != 0:
    errno = ctypes.get_errno()
    raise OSError(errno, f"prctl(PR_SET_PTRACER, ANY) failed: {ctypes.cast(libc.strerror(errno), ctypes.c_char_p).value.decode()}")
else:
    print("✅ Allowed ptrace from any same-UID process (PR_SET_PTRACER_ANY)")
```

it works in my environment.

Alternatively, it would be great if pytorch can support the `CU_MEM_HANDLE_TYPE_FABRIC` approach mentioned by @ngimel above, that requires cuda 12.5+ (and corresponding driver) though.
